rpc: introduce ContextOptions

tbg opened the pull request:
`rpc.NewContext` was one of these methods that could use one, especially
since it's about to pick up a TenantID.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50087)
<!-- Reviewable:end -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50087)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/context.go, line 645 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50087#-M9YZdrv58NjbNeCJFx2:-M9YZdrv58NjbNeCJFx3:b-r2rxcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be4bc5dcdc7911d99819eb6f4eff593f4cd3ffb/pkg/rpc/context.go#L645)):*
> ```Go
> 		dialOpts = append(dialOpts, grpc.WithInsecure())
> 	} else {
> 		tlsConfig, err := ctx.Config.GetClientTLSConfig()
> ```

@bdarnell Somewhat tangential to this PR, but this is part of the sin in our client cert handling. `base.Config.GetClientTLSConfig` uses the certificate manager, but I don't think we want that to be used for the various `cockroach` client commands. Perhaps a `ContextConfig.GetClientTLSConfig` hook could be added that is invoked if non-nil.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 27 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50087#-:-M9_46iuEOWVjJ_jCcbm:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/context.go, line 433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50087#-M9_3cHS9JYALRV3ZCHm:-M9_3cHS9JYALRV3ZCHn:b-dieh34) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be4bc5dcdc7911d99819eb6f4eff593f4cd3ffb/pkg/rpc/context.go#L433)):*
> ```Go
> // NewContext creates an rpc.Context with the supplied values.
> func NewContext(opts ContextOptions) *Context {
> 	if opts.Clock == nil {
> ```

How do you feel about adding a `ContextOptions.validate` method to enforce the "All pointer fields are required." requirement?


<!-- Sent from Reviewable.io -->

tbg committed:
rpc: introduce ContextOptions

`rpc.NewContext` was one of these methods that could use one, especially
since it's about to pick up a TenantID.

Release note: None

tbg commented:
TFTR! `validate` method added.

bors r=nvanbenschoten

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50087)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten, @petermattis, and @tbg)

___
*[pkg/rpc/context.go, line 645 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50087#-M9YZdrv58NjbNeCJFx2:-M9d0spY2087LBeDjKet:bri8hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be4bc5dcdc7911d99819eb6f4eff593f4cd3ffb/pkg/rpc/context.go#L645)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@bdarnell Somewhat tangential to this PR, but this is part of the sin in our client cert handling. `base.Config.GetClientTLSConfig` uses the certificate manager, but I don't think we want that to be used for the various `cockroach` client commands. Perhaps a `ContextConfig.GetClientTLSConfig` hook could be added that is invoked if non-nil.
</blockquote></details>

Yeah, I want to get base.Config out of the TLS cert business completely and pass things into the rpc.Context, so this is a nice step in that direction. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50087)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/rpc/context.go, line 645 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50087#-M9YZdrv58NjbNeCJFx2:-M9d3TohDDcOTt7l4T1l:bdxacmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be4bc5dcdc7911d99819eb6f4eff593f4cd3ffb/pkg/rpc/context.go#L645)):*
> Yeah, I want to get base.Config out of the TLS cert business completely and pass things into the rpc.Context, so this is a nice step in that direction.

+1 


<!-- Sent from Reviewable.io -->
